File: waveprop/waveprop.cpp
#include "waveprop.h"

// Initialize wave amplitude matrices
static float u_past[size][size];    // Wave amplitude at time n-1
static float u_present[size][size]; // Wave amplitude at time n
static float u_future[size][size];  // Wave amplitude at time n+1
static int step = 0;

void waveprop(hls::stream<packet>& output_stream)
{
    // Parameters
    float c = 343.0; // Speed of sound in air (m/s)
    float h = 1.0;   // Spatial step size (meters)

    // Stability condition (Courant–Friedrichs–Lewy condition)
    float safety_factor = 0.5;
    float delta_t = (h / c) * (1.0 / sqrt(2.0)) * safety_factor; // Time step size
    float s = c * delta_t / h;                                   // CFL number
    float s2 = s * s;

    // Initial condition: Gaussian pulse at the center
    float xc = (size - 1) * h / 2.0;
    float yc = (size - 1) * h / 2.0;
    float sigma = 5.0 * h;

    if (step == 0)
    {
    init_loop_i:
        for (int i = 0; i < size; ++i)
        {
#pragma HLS LOOP_TRIPCOUNT max = 100 avg = 100 min = 100
            float x = i * h;
        init_loop_j:
            for (int j = 0; j < size; ++j)
            {
#pragma HLS LOOP_TRIPCOUNT max = 100 avg = 100 min = 100
                float y = j * h;
                u_present(i, j) = hls::exp(-((x - xc) * (x - xc) + (y - yc) * (y - yc)) / (2.0 * sigma * sigma));
            }
        }
        u_past = u_present;
    }

update_loop_i:
    for (int i = 1; i < size - 1; ++i)
    {
#pragma HLS LOOP_TRIPCOUNT max = 100 avg = 100 min = 100
    update_loop_j:
        for (int j = 1; j < size - 1; ++j)
        {
#pragma HLS LOOP_TRIPCOUNT max = 100 avg = 100 min = 100
            u_future(i, j) = 2.0 * u_present(i, j) - u_past(i, j) +
                             s2 * (u_present(i + 1, j) + u_present(i - 1, j) +
                                   u_present(i, j + 1) + u_present(i, j - 1) -
                                   4.0 * u_present(i, j));


            packet output;
            output.data = u_future(i, j);
            output.keep = -1;
            output.strb = -1;
            //output.user = input_data.user;
            //output.last = input_data.last;
            //output.id = input_data.id;
            //output.dest = input_data.dest;

            if(i == size-2 && j == size - 2)
                output.last = true;
            else
                output.last = false;

            output_stream.write(output);
        }
    }
    // Apply boundary conditions (absorbing boundaries)
    for (int i = 0; i < size; ++i)
    {
#pragma HLS LOOP_TRIPCOUNT max = 100 avg = 100 min = 100
        u_future(i, 0) = 0.0;
        u_future(i, size - 1) = 0.0;
        u_future(0, i) = 0.0;
        u_future(size - 1, i) = 0.0;
    }

    // Update the wave amplitude matrices for the next iteration
    u_past = u_present;
    u_present = u_future;
}

File: waveprop/waveprop.py
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.animation as animation

# Parameters
c = 343.0  # Speed of sound in air (m/s)
h = 1.0    # Spatial step size (meters)
N = 100    # Number of spatial grid points along each axis
x = np.linspace(0, (N - 1) * h, N)
y = np.linspace(0, (N - 1) * h, N)
X, Y = np.meshgrid(x, y)

# Stability condition (Courant–Friedrichs–Lewy condition)
safety_factor = 0.5
delta_t = (h / c) * (1 / np.sqrt(2)) * safety_factor  # Time step size
s = c * delta_t / h  # CFL number
s2 = s * s

# Initialize wave amplitude arrays
u_past = np.zeros((N, N))     # Wave amplitude at time n-1
u_present = np.zeros((N, N))  # Wave amplitude at time n
u_future = np.zeros((N, N))   # Wave amplitude at time n+1

# Initial condition: Gaussian pulse at the center
xc = (N - 1) * h / 2
yc = (N - 1) * h / 2
sigma = 3 * h
u_present[:, :] = np.exp(-((X - xc) ** 2 + (Y - yc) ** 2) / (2 * sigma ** 2))
u_past[:, :] = u_present[:, :]

# Set up the plot
fig, ax = plt.subplots()
im = ax.imshow(u_present, vmin=-1, vmax=1, cmap='jet',
               extent=[0, (N - 1) * h, 0, (N - 1) * h])
fig.colorbar(im)
ax.set_title('2D Sound Propagation Simulation')
ax.set_xlabel('x (m)')
ax.set_ylabel('y (m)')

def update(frame):
    global u_present, u_past, u_future

    # Compute the wave amplitude at the next time step
    u_future[1:-1, 1:-1] = (2 * u_present[1:-1, 1:-1] - u_past[1:-1, 1:-1] +
                            s2 * (u_present[2:, 1:-1] + u_present[:-2, 1:-1] +
                                  u_present[1:-1, 2:] + u_present[1:-1, :-2] -
                                  4 * u_present[1:-1, 1:-1]))
    # Apply boundary conditions (absorbing boundaries)
    u_future[0, :] = u_future[-1, :] = u_future[:, 0] = u_future[:, -1] = 0

    # Update the wave amplitude arrays for the next iteration
    u_past, u_present = u_present, u_future.copy()

    # Update the plot
    im.set_array(u_present)
    return [im]

# Create the animation
ani = animation.FuncAnimation(fig, update, frames=200, interval=30, blit=True)
plt.show()
